Update the getting started guide (#97)
openrewrite/rewrite-docs: getting-started/getting-started.md

---

# Quick Start: Setting Up Your Project and Running Recipes

To help orient you to OpenRewrite, let's walk through configuring a project to use the Maven or Gradle rewrite plugin. Then let's walk through running various types of recipes on said project and talk through the results.

In this guide you will:

* [Clone a sample project](#step-1-clone-sample-project)
* [Add the rewrite-maven-plugin or rewrite-gradle-plugin to your project](#step-2-add-rewrite-maven-plugin-or-rewrite-gradle-plugin-to-your-project)
* [Activate a recipe so it can be run](#step-3-activate-a-recipe)
* [Run a simple refactoring recipe](#step-4-run-a-simple-refactoring-recipe)
* [Run a recipe with YAML configuration](#step-5-run-a-recipe-with-yaml-configuration)
* [Add and run an externally created recipe](#step-6-running-recipes-from-external-modules)

## Prerequisites

This quick start guide assumes that you:

* Are somewhat familiar with Java
* Have worked with a Maven or a Gradle project before
* Know how to run commands in the command line
* Can run basic Git commands

## Step 1: Clone sample project

> The first step in this process is making sure you have some code to work with. We've created a [sample repository](https://github.com/openrewrite/spring-petclinic-migration) that you can use if you'd like. However, as all of the steps in this guide apply to any Java project built with Maven or Gradle, please feel free to use your own and skip to [Step 2](#step-2-add-rewrite-maven-plugin-or-rewrite-gradle-plugin-to-your-project).

Reviewer: ```suggestion
The first step in this process is making sure you have some code to work with. We've prepared a [sample repository](https://github.com/openrewrite/spring-petclinic-migration) that you can use if you'd like. However, as all of the steps in this guide apply to any Java project built with Maven or Gradle, please feel free to use your own and skip to [Step 2](#step-2-add-rewrite-maven-plugin-or-rewrite-gradle-plugin-to-your-project).
```
I'm changing the word "created" to "prepared", because spring-petclinic was created by the spring team as a demo spring app. We forked that and made it suitable for our purposes.

{% endtab %}

{% tab title="Gradle" %}

* Add the rewrite plugin to the `plugins` section of your `build.gradle` file
* Make sure `mavenCentral()` is included in the `repositories` section
* Add a `rewrite` section that will be filled in later

Your file should look similar to:

{% code title="build.gradle" %}
```groovy
plugins {
    id 'java'
    id 'maven-publish'

>     id 'org.openrewrite.rewrite' version '5.32.0'

Reviewer: The changes you've made to omit parenthesis and prefer single-quoted to double-quoted strings _do_ make these more idiomatic groovy. But Gradle scripts can be written in Groovy or Kotlin. And Kotlin has no single-quoted string and does not allow parenthesis to be omitted in these circumstances.

But part of why I used double-quoted strings and included parenthesis is that the resulting code would, generally, work _either_ in groovy or kotlin scripts. If we're going to make the groovy in these code snippets more idiomatic, which is obviously not a bad thing, we might consider _also_ including gradle kotlin script examples.

Author: That's fair re the Kotlin side of things. The reason I switched this, though, is because when I checked out the sample repository that we suggest, it doesn't use the double quotes and parenthesis. I figure if someone was following along, it would be nicer to have the file look like it would in the sample repository. 

Would you mind explaining what you mean by `also including gradle kotlin script examples`? I'm not sure I follow that.

Author: Going to merge this for now but happy to update later once I better understand the ask :)

rewrite {
    // Will configure in subsequent steps
}

// ...
```
{% endcode %}

> 

Author: I didn't follow the hint below when I was reading through this guide but I left it in as I presumed it was important. If one of y'all could take a look and double-check (and perhaps provide me with some additional context) that would be swell!

```When applied to a multi-project build, behavior differs depending on whether the plugin is applied to the root project or to a sub-project. Applied to the root project, the plugin will parse and refactor all sources from all projects. Applied to any project other than the root project, the plugin will parse and refactor only sources from that project.```

Reviewer 2: I think this is really the expected behavior (of any gradle build task): 

if you run `./gradlew rewriteDryRun`, I think most developers would expect that to work on the entire project, whereas, `./gradlew :subProject:rewriteDryRun` would just work on the subproject. 

Not sure if there is something specific we are trying to call out here. I might ask @sambsnyd

Reviewer: That note could probably be mentioned only in the documentation on the gradle plugin itself, rather than here. Someone following our prescriptive guidance would not have to make any decision around this

Author: Alrighty -- I removed it to reduce confusion. Thanks!

{% tab title="Gradle" %}
{% code title="build.gradle" %}
```groovy
plugins {
    id 'java'
    id 'maven-publish'
    id 'org.openrewrite.rewrite' version '5.32.0'
}

rewrite {
    activeRecipe(
        'org.openrewrite.java.format.AutoFormat',
        'com.yourorg.VetToVeterinary'
    )
}
```
{% endcode %}
{% endtab %}
{% endtabs %}

> Once this recipe has been added to your active recipes, you can run either `./mvnw rewrite:run` or `./gradelw rewriteRun` to execute all of your active recipes. Afterward, you'll see that:

Reviewer: ```suggestion
Once this recipe has been added to your active recipes, you can run either `./mvnw rewrite:run` or `./gradlew rewriteRun` to execute all of your active recipes. Afterward, you'll see that:
```

{% hint style="info" %}
You can search through all of the recipes in the [OpenRewrite docs](https://docs.openrewrite.org/reference/recipes). Each recipe page has instructions for how to import the recipe and what parameters (if any) need to be included. 
{% endhint %}

Let's pretend that you want to migrate JUnit 4 to JUnit 5 in a Spring project you have. If you take a look at the [Usage section](https://docs.openrewrite.org/reference/recipes/java/spring/boot2/springboot2junit4to5migration#usage) in the [JUnit 4 to 5 migration recipe](https://docs.openrewrite.org/reference/recipes/java/spring/boot2/springboot2junit4to5migration), you'll see what you need to include in your `build.gradle` or `pom.xml` file in order to use this recipe. 

Below, we'll walk through the [Maven](#maven--external-modules) and [Gradle](#gradle--external-modules) changes and provide some additional context around said changes.

### Maven + External Modules

For Maven projects, you'll need to: 

* Add the recipe to the `activeRecipes` list
* Add a dependency on `rewrite-spring` (where the JUnit 4 to 5 recipe lives)

> * Specify a version of `rewrite-spring` to use

Author: I'm unsure how someone would know which version to specify for a given recipe. It seems like there's a possibility that people could specify the wrong version or have versions that are mismatched. Is that true? Could anyone give me some additional context on that front?

Reviewer 2: Well, we actually specify the version in the example pom.xml. We update that version with each release.

<build>
```
{% endcode %}

To double-check that everything is working, run the command `./mvnw rewrite:run`. Your project should be upgraded to Spring Boot 2 and all of the test classes should be updated to JUnit 5. Your `pom.xml` file will also have had its Spring dependencies updated, the JUnit 4 dependency removed, and the JUnit 5 dependency added.

### Gradle + External Modules

Unlike Maven projects, Gradle projects have two options for specifying versions for recipes. You can:

> 1. Add `rewrite-recipe-bom` as a dependency

Reviewer: ```suggestion
1. Add `rewrite-recipe-bom` as a [bill of materials (BOM) dependency](https://docs.gradle.org/current/userguide/platforms.html#sub:bom_import)
```

Adding rewrite-recipe-bom as a regular dependency is not the same as using it to manage the versions of other dependencies. You accurately explain the difference below, but I don't want a discrepancy to cause confusion.

Author: Good call out and appreciate the suggested change.

2. Add `rewrite-spring` as a dependency with a version specified

If you choose to use the `rewrite-recipe-bom`, you won't have to worry about specifying versions for your rewrite recipes as all of the recipes you include in your `dependencies` section will have an appropriate version specified in the bill of materials (BOM). For Gradle projects, this is the recommended approach.

> If you choose to not use the `rewrite-recipe-bom`, you'll need to ensure that you have the appropriate version of all of your dependencies included (as you would with Maven projects).

Reviewer: ```suggestion
If you choose to not use `rewrite-recipe-bom`, you'll need to specify the version of each rewrite recipe module you use. 
```

Author: That's a great call out. Agree with this suggestion. Thanks
